melisma line adjustment for manual position & voice offset

MarcSabatella opened the pull request:
Take manual positioning (of the note) and offsets generated for multiple voices into consideration when calculating melisma line length.  Also moved the adjustment to add a constant amount of extra padding into SLine::linePos() where the rest of these calculations are taking place.  Result tested to be consistent for all note head types, stems up & down.

MarcSabatella commented:
@mgavioli : sorry, our PR's crossed.  I found through much experimentation that looping through notes is only way to get reliable figure for the right-hand edge of the chord.  But now that I see your PR, I think I see why your additional adjustment is in lyrics.cpp (so you can do it differently according to situation).  I guess we should merge these somehow.  Sorry for the confusion.

mgavioli commented:
@MarcSabatella : I assume you had good reasons for looping through the chord notes, but I can't find a case where the current code gives wrong results for different note values and/or up/down stems. Can you describe one?

There are problems with multi-voice, though; for instance this:

![test_lyrics_2_bis](https://cloud.githubusercontent.com/assets/1420743/5732195/23b838f4-9b8e-11e4-8d00-1e2fefe3ae5e.png)

Here the problem is with func `Score::findCRinStaff()` (called by `Spanner::computeEndElement()` called by `SLine::layout()`), which returns the whole measure rest in voice 1 rather than the semibreve chord in voice 2.

MarcSabatella commented:
Before either of our current PR's, the lengths were OK unless a note was offset because of a conflict with another voice (a second or overlap).  And I couldn't fix that without introducing inconsistencies, until I took note offsets into account.  I need to try your code, I guess.

The issue with findCRinStaff() also explains the voice bug I filed.  It seems other spanners deliberately cheat their lengths.  tick2 is set according to the tick of the next segment, not the end of the note.  I guess that is what lyric lines will need to do too.

MarcSabatella commented:
@mgavioli : I had a moment to check something out, and I believe that we can fix the multivoice issues by simplifying the code at the top of LyricsLine::layout() (I *love* it when that happens :-).  We don't need to go searching for a CR and add in its duration.  We just need to set the tick to the tick of the *next* segment, minus 1.  At least, I believe that's how other spanners do it.  I was lazy, and actually tried something simpler - I just called setTick() with the tick of *this* segment, plus 1.  Anything that puts the tick count somewhere clearly within the bounds of this segment would probably to do the trick, although for consistency we should probably do it the same way as other spanners.  I might have some time later tonight to play with this more.

MarcSabatella commented:
I have updated this to incorporate changes from @mgavioli in #1633.  That is, the fix to continuous mode, the change to shorten rather than lengthen non-ending melisma lines by minNoteDistance.  I also added fixes for the voice issues - see http://musescore.org/en/node/44271.  Everything checks out from what I can tell, although there are still a couple of bugs left to fix.

MarcSabatella commented:
My code here really shows bad problems with http://musescore.org/en/node/44291.  As I said, I think there are bugs there anyhow, but my code does worse things than before as a result, I think.  I will need to look at that tomorrow.

mgavioli commented:
It does not really make sense to step on each other feet ;) .

So, I am leaving this area to you. I noticed there are weaknesses in the melisma entering process (mainly `ScoreView::lyricsUnderscore()` in `editlyrics.cpp`) and I'll rather have a look at them.

MarcSabatella commented:
I am sorry, I know I should not have rushed in.  But I was excited :-).  Anyhow, I'm also excited to see you are working on the  process of entering the melismas!

MarcSabatella commented:
Closing while I deal with http://musescore.org/en/node/44291.  I have it mostly working, hope to update later today.

MarcSabatella commented:
This should fix a number of issues:

- manual positioning of notes, as well as offsets due to voice collision, taken into consideration

- continuous view layout (courtesy of @mgavioli)

- shortening of start lyrics line segments to end just before barline (also courtesy of @mgavioli)

- fix for multivoice issues (see http://musescore.org/en/node/44271)

- fix so that edits to notes / measures don't leave lyrics with invalid tick counts - melisma is adjusted to nearest chord (see http://musescore.org/en/node/44291)

There might be a manual rebase required relative to #1636, as they come *close* to touching the same code.

MarcSabatella commented:
Looks like I need to re-submit this later to re-open it.
